File: backend.core/pkg/bible/usx/char.go
/*
 * Copyright 2024 Sowers, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package usx

import (
	"encoding/xml"
	"errors"
	"fmt"
)

type Char struct {
	Text string
}

func (c *Char) AddText(text string) error {
	c.Text += text
	return nil
}

func (p *Parser) onChar(element xml.StartElement, current NodeContainer) (bool, error) {
	char := &Char{}
	for _, attr := range element.Attr {
		switch attr.Name.Local {
		case "style":
			//char.Style = parseCharStyle(attr)
		}
	}
	for {
		token, err := p.decoder.Token()
		if err != nil {
			return false, err
		}
		switch element := token.(type) {
		case xml.CharData:
			char.Text = string(element)
		case xml.EndElement:
			if element.Name.Local != "char" {
				return false, errors.New("onChar: unexpected end")
			}
			current.AddNode(char)
			return true, nil
		default:
			return false, fmt.Errorf("onChar: unexpected element: %v", element)
		}
	}
}

File: backend.workflows/packages/bible/src/usx/char.ts
/*
 * Copyright 2024 Sowers, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { CharStyle, CharStyles } from './styles'
import { Attributes, StyleFactoryFilter, UsxContext, UsxItemContainer, UsxItemFactory } from './item'
import { Footnote, FootnoteFactory } from './footnote'
import { Break, BreakFactory } from './break'
import { Text, TextFactory } from './text'
import { Reference, ReferenceFactory } from './reference'
import { Milestone, MilestoneFactory } from './milestone'

type CharType = Reference | Char | Milestone | Footnote | Break | Text

export class Char extends UsxItemContainer<CharType> {

  style: CharStyle
  //char.link?,
  //char.closed?,

  constructor(context: UsxContext, attributes: Attributes) {
    super(context, attributes)
    this.style = attributes.STYLE.toString() as CharStyle
  }
}

export class CharFactory extends UsxItemFactory<Char> {

  static readonly instance = new CharFactory()

  private constructor() {
    super('char', new StyleFactoryFilter(CharStyles))
  }

  protected onInitialize() {
    this.register(ReferenceFactory.instance)
    this.register(CharFactory.instance)
    this.register(MilestoneFactory.instance)
    this.register(FootnoteFactory.instance)
    this.register(BreakFactory.instance)
    this.register(TextFactory.instance)
  }

  create(context: UsxContext, attributes: Attributes): Char {
    return new Char(context, attributes)
  }
}
